### PR TITLE
support pip download with downloaded state option

### PR DESCRIPTION
**Pull request by justinsousa**
##### SUMMARY

I often use the pip download command to download python libraries that are dependencies for projects. I figure adding a `downloaded` state which essentially just calls pip download would make sense.  It would allow downloading all dependencies for a project using the pip module and then later pointing pip at that directory with extra_args containing  -f/--find-links pointing at the directory that the dependencies were downloaded to. This would avoid the need to call pip download with the command/shell module.

Some example uses below

```
- name: download library by name
  become: no
  pip:
    chdir: /myapp/pylibs
    name: Bottle
    state: downloaded
    executable: /usr/local/bin/pip2.7

- name: download from reqs file
  become: no
  pip:
    chdir: "{{ pylib_cache }}"
    requirements: /myapp/requirements.txt
    state: downloaded
```

Something to note, the download option was added with version 8.0 (I believe that's the version based on https://pip.pypa.io/en/stable/news/#id18), so it might be necessary to parse the version of pip and change the command to the old style `install --download` below that version.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

pip module

##### ANSIBLE VERSION

```
ansible 2.8.0.dev0
  config file = None
  configured module search path = ['${HOME}/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = ${HOME}/repos/py3venvs/ans-dev/lib/python3.7/site-packages/ansible-2.8.0.dev0-py3.7.egg/ansible
  executable location = ${HOME}/repos/py3venvs/ans-dev/bin/ansible
  python version = 3.7.0 (default, Jul 23 2018, 20:24:19) [Clang 9.0.0 (clang-900.0.39.2)]

```

### CODE REVIEW

**Comment by robinro** [comment]
Thanks @justinsousa for this PR.
Overall I think this feature is a good addition to the pip module.

Could you please add tests? See `test/integration/targets/pip/tasks/pip.yml`.
The tests should especially cover
- correctness (file foo.tar.gz or foo.whl in current directory)
- idempotency: first run changed is True, second changes is False
- reasonable defaults (works without extra_args and then uses the cwd for output)

**Comment by robinro** [comment]
As far as I see the test failures are unrelated to this change.

**Comment by justinsousa** [comment]
sure thing. looks like, as you pointed out, the tests that are failing don't appear related. i'lll check it out nonetheless and add some others. right now, running the command twice still reports the task as changed so I'll need to dig into that anyways. thanks @robinro

**Comment by robinro** [comment]
The failing tests will be covered by https://github.com/ansible/ansible/pull/39460

**Comment by robinro** [comment]
@justinsousa Please let me know if you want help with this.

**Comment by justinsousa** [comment]
@robinro will do...just been a busy week. looking to finish up tomorrow

**Comment by justinsousa** [comment]
ah the tests I wrote are invalid given the order they are run right now. will need to revisit how to test it

I wanted to test that the first time a package is downloaded, the source code / wheel is actually retrieved and saved and then the next time you attempt to run pip download with the same dependency (in the same location) you would not see "Saved" in the output. I thought this would be a good way to determine whether the status should be marked as changed. 

I was piggy backing off of the test_package which of course is already downloaded by the time my test ran. might be a good place to combine chdir test and downloading it to a different location rather than the default (/tmp) as the other tests are downloading there.

```
2018-05-01 03:04:05     "stdout_lines": [
2018-05-01 03:04:05         "Collecting tex", 
2018-05-01 03:04:05         "  File was already downloaded /tmp/tex-1.8.tar.gz", 
2018-05-01 03:04:05         "Successfully downloaded tex"
```
The other thing that's a bit strange about the pip download output is that it reports `Successfully downloaded` even if the call being run didn't actually download the dependency (because it is already downloaded). Hence, the cleanest way I saw to determine that something happened that changed the state was parsing all of the command output for "Saved". For example, a reqs file that results in 5 libs which already downloaded but 1 that is new would result in an output that would contain 6 "Successfully downloaded" messages and 1 "Saved".

File: lib/ansible/modules/packaging/language/pip.py [hunk 1]
```diff
@@ -559,6 +571,7 @@ def main():
         absent=['uninstall', '-y'],
         latest=['install', '-U'],
         forcereinstall=['install', '-U', '--force-reinstall'],
+        downloaded=['download'],
```
Comment (justinsousa) [inline lib/ansible/modules/packaging/language/pip.py]: @robinro figured I'd revive this PR. notice that some of the tests that are failing are due to what I pointed out in the main PR comment that the download option is a newer one added to pip not too long ago. where would you prefer I check that download is an option? Probably makes the most sense to allow pip to be found first, then check if download is an option and fail if not.

**Pull request closed without merging**